native/src/widget/grid.rs

```diff
@@ -0,0 +1,198 @@
+///
+/// # Example
+///
+/// ```
+/// # use iced_native::{renderer::Null, Element, Grid as NativeGrid, Text};
+/// # type Grid<'a> = NativeGrid<'a, (), Null>;
+/// Grid::from_vec(
+///     2,
+///     vec![
+///         Element::from(Text::new("First row, first column")),
+///         Element::from(Text::new("First row, second column")),
+///         Element::from(Text::new("Second row, first column")),
+///         Element::from(Text::new("Second row, second column")),
+///     ],
+/// );
+/// ```
+#[allow(missing_debug_implementations)]
+pub struct Grid<'a, Message, Renderer> {
+    columns: usize,
+    elements: Vec<Element<'a, Message, Renderer>>,
+}
+
+impl<'a, Message, Renderer> Grid<'a, Message, Renderer> {
+    /// Create a new empty [`Grid`].
+    ///
+    /// [`Grid`]: struct.Grid.html
+    pub fn new(columns: usize) -> Self {
+        Self::from_vec(columns, Vec::new())
+    }
+
+    /// Create a new [`Grid`] with the given [`Element`]s.
+    ///
+    /// [`Grid`]: struct.Grid.html
+    /// [`Element`]: ../struct.Element.html
+    pub fn from_vec(
+        columns: usize,
+        elements: Vec<Element<'a, Message, Renderer>>,
+    ) -> Self {
+        Self { columns, elements }
+    }
+
+    /// Adds an [`Element`] to the [`Grid`].
+    ///
+    /// [`Element`]: ../struct.Element.html
+    /// [`Grid`]: struct.Grid.html
+    pub fn push<E>(mut self, element: E) -> Self
+    where
+        E: Into<Element<'a, Message, Renderer>>,
+    {
+        self.elements.push(element.into());
+        self
+    }
+
+    /// Adds [`Element`]s to the [`Grid`].
+    ///
+    /// [`Element`]: ../struct.Element.html
+    /// [`Grid`]: struct.Grid.html
+    pub fn push_vec(
+        mut self,
+        mut elements: Vec<Element<'a, Message, Renderer>>,
+    ) -> Self {
+        self.elements.append(&mut elements);
+        self
+    }
+}
+
+impl<'a, Message, Renderer> Widget<Message, Renderer>
+    for Grid<'a, Message, Renderer>
+where
+    Renderer: self::Renderer,
+{
+    fn width(&self) -> Length {
+        Length::Shrink
+    }
+
+    fn height(&self) -> Length {
+        Length::Shrink
+    }
+
+    fn layout(
+        &self,
+        renderer: &Renderer,
+        limits: &layout::Limits,
+    ) -> layout::Node {
+        // store calculated layout sizes
+        let mut layouts = Vec::with_capacity(self.elements.len());
+        // store width of each column
+        let mut column_widths = Vec::<f32>::with_capacity(self.columns);
+
+        // find out how wide a column is by finding the widest cell in it
+        for (column, element) in (0..self.columns).cycle().zip(&self.elements) {
+            let layout = element.layout(renderer, limits).size();
+            layouts.push(layout);
+
+            if let Some(column_width) = column_widths.get_mut(column) {
+                *column_width = column_width.max(layout.width);
+            } else {
+                column_widths.insert(column, layout.width);
+            }
+        }
```

Comment:
Looks cool! Gotta love iterators.

I think it'd be interesting to also be able to provide a fixed column width on creation and then compute the number of columns based on the `limits` instead of the children. In this case, the `Grid` would have `Fill` width behavior instead.

Reply:
Should I remove fixed column then and just take column width?
If we want both to be possible, how would we approach that:
-  pass an enum?
-  make two different `new_*` functions?
-  what should be the feault?

I don't particularly understand the use case here, is it like an image gallery? Can you give me some examples? It would also help me make the actual grid example, because I was thinking of something table-like, which is, obviously, more table then grid.

Reply:
Looking at the other widgets I figured we could just do `fn columns(self, columns: usize) -> Self`, basically following the builder pattern here.

Reply:
I hope this is better now.

Reply:
So I made it like this:
-  **Defined only column:** determine column width by widest element in that column
-  **Defined only column width:** determine number of columns by dividing max width by column width
-  **Defined both:** number of columns and column width is respected as expected
-  **Defined none:** elements will be displayed exactly as in row

Anything we want to tweak here?